File: Exercism/haskell/list-ops/src/ListOps.hs
{-# LANGUAGE BangPatterns #-}

module ListOps
  ( length
  , reverse
  , map
  , filter
  , foldr
  , foldl'
  , (++)
  , concat
  ) where

import Prelude hiding
  ( length, reverse, map, filter, foldr, (++), concat )

foldl' :: (b -> a -> b) -> b -> [a] -> b
foldl' _ z [] = z
foldl' f z (x:xs) = x `seq` foldl' f (f z x) xs

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr _ z [] = z
foldr f z (x:xs) = f x (foldr f z xs)

length :: [a] -> Int
length = foldr (\_ acc -> acc + 1) 0

reverse :: [a] -> [a]
reverse = foldl (flip (:)) []

map :: (a -> b) -> [a] -> [b]
map f = foldr (\x acc -> f x : acc) []

filter :: (a -> Bool) -> [a] -> [a]
filter p = foldr _select []
  where _select x acc | p x       = x:acc
                      | otherwise = acc
(++) :: [a] -> [a] -> [a]
xs ++ ys = foldr (:) ys xs

concat :: [[a]] -> [a]
concat = foldr (++) []

File: Exercism/python/list-ops/list_ops.py
from operator import add
from functools import reduce


def compose2(fn1, fn2):
    return lambda *args: fn1(fn2(*args))


def flip(fn):
    return lambda a, b: fn(b, a)


append = add
filter = compose2(list, __builtins__["filter"])
length = len
map = compose2(list, __builtins__["map"])
foldl = reduce
reverse = compose2(list, reversed)


def concat(lists):
    return reduce(add, lists, [])


def foldr(f, list_, initial):
    return reduce(flip(f), reversed(list_), initial)
